server/lorax_server/models/custom_modeling/attentions/self_extend/mistral_self_extend_patch.py

```diff
@@ -0,0 +1,190 @@
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
+    # The first two dimensions of cos and sin are always 1, so we can `squeeze` them.
+    cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
+    sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
+    cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    q_embed = (q * cos[:,:, -q.shape[2]:]) + (rotate_half(q) * sin[:,:, -q.shape[2]:]) if q is not None else None
+    k_embed = (k * cos) + (rotate_half(k) * sin) if k is not None else None
+    return q_embed, k_embed
+
+def apply_grouped_rotary_pos_emb(q, k, cos, sin, position_ids, g_size_1=1, g_size_2=4096):
+    # The first two dimensions of cos and sin are always 1, so we can `squeeze` them.
+    position_ids_q = position_ids//g_size_1 + g_size_2 - g_size_2//g_size_1
+    position_ids_k = position_ids//g_size_1
+
+    cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
+    sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
+    cos_q = cos[position_ids_q].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    sin_q = sin[position_ids_q].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    cos_k = cos[position_ids_k].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    sin_k = sin[position_ids_k].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    q_embed = (q * cos_q) + (rotate_half(q) * sin_q) if q is not None else None
+    k_embed = (k * cos_k) + (rotate_half(k) * sin_k) if k is not None else None
+
+    return q_embed, k_embed
+
+def apply_neighbor_rotary_pos_emb(q, k, cos, sin, position_ids, g_size=1):
+    # The first two dimensions of cos and sin are always 1, so we can `squeeze` them.
+    position_ids = position_ids % g_size
+
+    cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
+    sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
+    cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+def apply_identical_rotary_pos_emb(q, k, cos, sin, position_ids, idd_position=1024):
+    position_ids = torch.ones_like(position_ids) * idd_position
+
+    cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
+    sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
+    cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+def self_extend_forward(

```

Comment:
This signature is quite different from what the current implementation without self-extend uses. Instead of using flash attention and paged attention, it uses the more conventional attention computation with `past_key_values`. As such, because the rest of the FlashMistral, etc. classes don't pass in the `past_key_values`, my expectation is that the attention computation will be incorrect during the decode phase.

I think what we need is a variation on this function that works with the existing Flash / Paged Attention computation.

Reply:
I tried with and without the self extend and the generated response was the same
The actual implementation is used from the authors, i think (https://github.com/datamllab/LongLM)

I definitely think that we should add the flash attention version, but dont know if it would make sense to wait until they released it with tested results

from the paper:

> Limitation: The limitation of the proposed Self-Extend includes the lack of implementation of Flash Attention (Dao
> et al., 2022) and the performance degradation with too large
> group size, which means the context window still cannot be
> extended to infinity with current SelfExtend. Meanwhile,
> like many regular tasks, there is still no consensus at present
> about how to do evaluation for long context tasks, which
> may cause problematic evaluation results.


> Future Work: For future work, we will implement Flash
> Attention for Self-Extend to enhance its efficiency. We
> are also interested in testing SelfExtend on models using
> other positional encoding. Larger models, longer context
> and more challenging tasks will be tested if we can have
> access to more computational resources in the future. In
> the meantime, more sophisticated mapping methods will
> be considered as the replacement of the simple FLOOR operation, so as to achieve better long context understanding
> abilities and longer extended context window length.

Reply:
Hey @flozi00, it looks like there was a bug in the code causing the self extend code path to never be executed. I fixed the issue with plumbing through the `self_extend_attention` param, and now there are some errors showing up. So my suspicion is that the answers were identical because we were executing the non-extended code in both cases.

Here's the current error:

```
  File "/data/lorax/server/lorax_server/models/flash_mistral.py", line 427, in forward                                                                                            
    logits = model.forward(                                                                                                                                                       
  File "/data/lorax/server/lorax_server/models/custom_modeling/flash_mistral_modeling.py", line 565, in forward                                                                   
    hidden_states = self.model(                                                                                                                                                   
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl                                                                     
    return self._call_impl(*args, **kwargs)                                                                                                                                       
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl                                                                             
    return forward_call(*args, **kwargs)                                                                                                                                          
  File "/data/lorax/server/lorax_server/models/custom_modeling/flash_mistral_modeling.py", line 508, in forward                                                                   
    hidden_states, residual = layer(                                                                                                                                              
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl                                                                     
    return self._call_impl(*args, **kwargs)                                                                                                                                       
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl                                                                             
    return forward_call(*args, **kwargs)                                                                                                                                          
  File "/data/lorax/server/lorax_server/models/custom_modeling/flash_mistral_modeling.py", line 431, in forward                                                                   
    attn_output = self.self_attn(                                                                                                                                                 
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl                                                                     
    return self._call_impl(*args, **kwargs)                                                                                                                                       
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl                                                                             
    return forward_call(*args, **kwargs)                                                                                                                                          
TypeError: self_extend_forward() got multiple values for argument 'group_size_1' 
```

Reply:
oh, okay
will take this tomorrow
Thanks for finding this missing one